fx quant: do not observe bias on F.conv

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #49740 fx quant: split linear test cases
* #49717 fx quant: hook up ConvTranspose{n}d
* #49688 fx quant: improve types on convert
* #49628 fx quant: do not observe bias on F.linear
* **#49623 fx quant: do not observe bias on F.conv**

Summary:

Ensures that conv bias is not observed in a `F.conv{n}d` call.  This should be a small speedup in PTQ, and will change numerics (in a good way) for QAT if someone is using `F.conv`.

Test Plan:

```
python test/test_quantization.py TestQuantizeFxOps.test_conv_functional_bias_not_observed
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25652856](https://our.internmc.facebook.com/intern/diff/D25652856)

vkuzo committed:
fx quant: do not observe bias on F.conv

Summary:

Ensures that conv bias is not observed in a `F.conv{n}d` call.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
nit on "fx quant: do not observe bias on F.conv"


Summary:

Ensures that conv bias is not observed in a `F.conv{n}d` call.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "fx quant: do not observe bias on F.conv"


Summary:

(not ready for review)

Ensures that conv bias is not observed in a `F.conv{n}d` call.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "fx quant: do not observe bias on F.conv"


Summary:

(not ready for review)

Ensures that conv bias is not observed in a `F.conv{n}d` call.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25652856](https://our.internmc.facebook.com/intern/diff/D25652856)

[ghstack-poisoned]

vkuzo committed:
Update on "fx quant: do not observe bias on F.conv"


Summary:

Ensures that conv bias is not observed in a `F.conv{n}d` call.  This should be a small speedup in PTQ, and will change numerics (in a good way) for QAT if someone is using `F.conv`.

Test Plan:

```
python test/test_quantization.py TestQuantizeFxOps.test_conv_functional_bias_not_observed
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25652856](https://our.internmc.facebook.com/intern/diff/D25652856)

[ghstack-poisoned]

jerryzh168 commented on torch/quantization/fx/quantize.py:
```diff
@@ -237,10 +237,38 @@ def insert_observer_for_input_arg_of_observed_node(
+    torch.nn.functional.conv3d : [2],
+}
+
+def node_arg_is_bias(node: Node, arg: Any) -> bool:
```
nit: shall we share the code of this function with node_arg_is_weight as well

vkuzo replied:
I don't feel too strongly.  The definition diverges in the next PR though so not sure if worth it.

vkuzo replied:
lmk if a change is needed here given the context from the next PR, I'm flexible

jerryzh168 replied:
sounds good, we can keep the current approach

jerryzh168 commented on torch/quantization/fx/quantize.py:
```diff
@@ -958,15 +986,19 @@ def _find_quants(self, graph: Graph, matches: Dict[str, MatchResult],
+                        is_weight
+                    ) and
+                    # bias should not be observed
+                    (not is_bias)
```
I feel the condition should probably be:
```
qconfig is not None and ((activation_is_statically_quantized(qconfig) and not is_bias) or (activation_is_statically_quantized(qconfig) and is_weight)
```

# we didn't check the weight should be statically quantized before actually

for this PR I think we can merge (not is_bias) check with activation_is_statically_quantized check and add the weight check in a separate PR.

Basically
1. `(activation_is_statically_quantized(qconfig) and not is_bias` checks if the activation needs to be observed and we'll use qconfig.activation as act_post_process_ctr
2. `(activation_is_statically_quantized(qconfig) and is_weight)` checks if the weight needs to be observed and we'll use qconfig.weight as act_post_process_ctr

vkuzo replied:
I agree, what I have now is not the most readable.  Thinking about it some more, activation / weight / bias seem mutually exclusive, so maybe it should be 

```
is_activation = not (is_weight or is_bias)
(qconfig is not None and (
  (is_activation and activation_is_statically_quantized(qconfig)) or
  (is_weight and weight_is_statically_quantized(qconfig))
)
```
?

jerryzh168 replied:
sure, this sounds good

jerryzh168 commented on test/quantization/test_quantize_fx.py:
```diff
@@ -1330,9 +1330,23 @@ def forward(self, x):
                 quantized_nodes[dim])
 
+    @skipIfNoFBGEMM
+    def test_conv_functional_bias_not_observed(self):
```
optional: can we merge this test with previous test and call it "test_conv"? we have "test_linear" that checks all combinations of is_module/is_not_module and has_rleu/doesn't have relu and relu module/functional. I think that would be a more exhaustive testing.

vkuzo replied:
hmm, that was intentional, there are benefits when each test case is separate.  This allows for faster debugging when chasing down an individual issue (so there is no need to comment out the unrelated test cases in the large loop).  Perhaps we can align on a way to both share code and also test one thing at a time, lmk your preference.

jerryzh168 replied:
I think we should do exhaustive tests on what we want to support first. in this particular case, the problem is that we did not test functional conv before. Testing for special cases that we encounter in model might be too ad hoc and it will be hard to know what is covered and what is not in the future.

I think we can first make sure our unit test cover all possible modules/ops etc. we want to support, and then if there is any corner case we can cover with these ad hoc tests (typically should be some interaction between ops + qconfig instead of one op). one such example is here: https://github.com/pytorch/pytorch/blob/master/test/quantization/test_quantize_fx.py#L538

jerryzh168 replied:
regarding separating test cases I think as long as there is no code duplication it should be OK.

vkuzo replied:
sure, I made the functional test more comprehensive.

Regarding no code duplication, I think it's good to optimize for in general, but code readability and ease of use is something to consider as well.  Some of the current test utils seem to bias on the "minimize code reuse" side while sacrificing readability, at least for people who were not the code authors.  Maybe in the future we can consider on ways to do both.

jerryzh168 replied:
sure sounds good

vkuzo committed:
Update on "fx quant: do not observe bias on F.conv"


Summary:

Ensures that conv bias is not observed in a `F.conv{n}d` call.  This should be a small speedup in PTQ, and will change numerics (in a good way) for QAT if someone is using `F.conv`.

Test Plan:

```
python test/test_quantization.py TestQuantizeFxOps.test_conv_functional_bias_not_observed
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25652856](https://our.internmc.facebook.com/intern/diff/D25652856)

[ghstack-poisoned]

vkuzo committed:
Update on "fx quant: do not observe bias on F.conv"


Summary:

Ensures that conv bias is not observed in a `F.conv{n}d` call.  This should be a small speedup in PTQ, and will change numerics (in a good way) for QAT if someone is using `F.conv`.

Test Plan:

```
python test/test_quantization.py TestQuantizeFxOps.test_conv_functional_bias_not_observed
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25652856](https://our.internmc.facebook.com/intern/diff/D25652856)

[ghstack-poisoned]
